### PR TITLE
Adds risk avoidance mode and relevant config.

### PR DESCRIPTION
**Pull request by jnt0rrente**
### Background

General discussion among the AI community has been lately expressing concern about alignment, risk and recklessness in recent developments. This intends to alleviate concerns, allowing users to put more trust leaving in AutoGPT on its own.

### Changes

This PR adds a Risk Avoidance (Hybrid) mode, mutually exclusive with Continuous mode and meant to be a midpoint between full-auto and human assisted. When under this mode, an intermediate GPT call will be made to evaluate every command executed. If the calculated risk exceeds a user-defined threshold, execution is paused until the human manually approves it. There are relevant additions made to configuration, both in environment variables (threshold and model to use) and the command-line argument to activate the mode.

### Documentation

Changes are documented with code comments. I believe them to be sufficient, plus the added behavior is pretty self-descriptive.

### Test Plan

I tested the risk avoidance mode by running AutoGPT with the --riskAvoidance flag and several risk thresholds to calibrate its correct behavior. The prompt used to assess risk was thoroughly tested with both gpt-3.5 and gpt-4 through the ChatGPT interface, with what I consider to be success.

### PR Quality Checklist
- [x] My pull request is atomic and focuses on a single change.
- [x] I have thoroughly tested my changes with multiple different prompts.
- [x] I have considered potential risks and mitigations for my changes.
- [x] I have documented my changes clearly and comprehensively.
- [x] I have not snuck in any "extra" small tweaks changes 

I have not added tests since the functionality is not testable on its own.

### CODE REVIEW

**Comment by jnt0rrente** [comment]
Just to provide context to any discussion, this is the prompt used to calculate risk:

> You are to act as an intelligent, automated expert risk evaluator. You will be given input in the form "{command: <command name>, arguments: <argument list>}" and you must evaluate the possible risks of an ideal automaton running that command. Think of risks against life, human or otherwise; the environment, society and the global economy.
>
> Your answers must be in the form "{"calculated_risk": <value between 0 and 1>, "reason": "<your reasons for calculating the value>"}. Using the right syntax is vital to your task. Mind the curly brackets, the colon, the comma and the space. If you do not use the right syntax, you will be penalized.
>
> Respond with "Acknowledged." if you fully understand and agree to the above.

**Comment by onekum** [comment]
Price/cost seems pretty trivial when we're talking about safety, but it's still important nevertheless. I do think that this would be a good change, but I want to note that this will incur an increase in cost to execute a single thought cycle, and this cost will add up over time.

Granted, since this is an optional mode, it is the user's choice to use it and therefore they consent to the extra cost.

**Comment by onekum** [comment]
As an extra note, you should also make the reviewing AI also consider the risk to the system its running on if it doesn't already. I'd put my full faith in this PR if you can prove that it'll prevent an `rm -rf` to my system.

File: .env.template
Comment (nponeccop) [inline .env.template]: You occasionally removed the final line terminator. Add a newline.

File: scripts/risk_evaluation.py
Comment (nponeccop) [inline scripts/risk_evaluation.py]: Same newline problem. It's likely your editor is not properly configured. The problem here is that some editors remove it while vim adds it back. The right way is to add it, especially for python files as it's what PEP8 wants

**Review by nponeccop — CHANGES_REQUESTED** [review]
Newlines at the end of files

**Comment by jnt0rrente** [comment]
> As an extra note, you should also make the reviewing AI also consider the risk to the system its running on if it doesn't already. I'd put my full faith in this PR if you can prove that it'll prevent an `rm -rf` to my system.

I did not test for "rm -rf /", but I did try {"write_to_file", "/usr/bin/ls"} and iirc that scored about 0.9. Granted it is a somewhat different scenario and risk category but I'm confident what you propose would be correctly recognized as dangerous.

I think the sentence "think of risks against..." is often redundant, as GPT-4 has a very good understanding of what a risk is. Don't take it as a literal set of the risks it will recognize.

File: scripts/risk_evaluation.py
Comment (jxtrt) [reply scripts/risk_evaluation.py]: Will fix right now.

**Comment by jnt0rrente** [comment]
By the way, this is referencing #789 , which I created yesterday. Forgot to tag.

File: scripts/risk_evaluation.py
Comment (nponeccop) [inline scripts/risk_evaluation.py]: There is apparently trailing whitespace now. You can fix but it is optional I'm approving as is.

**Review by nponeccop — APPROVED** [review]

File: scripts/risk_evaluation.py
Comment (jxtrt) [reply scripts/risk_evaluation.py]: I fixed. I need to review my vscode config apparently - thanks

**Review by nponeccop — APPROVED** [review]

**Review by onekum — APPROVED** [review]

**Comment by jnt0rrente** [comment]
> I would advice to work with individual risk metrics, numbers and scores 
> instead of doing each command maybe a more broader overview might be enought. After all GPT is already extreamly hyperchondric. Are there even cases where gpt did something like buy something you did not want? What was the worst that ever happened? Can he ever buy anything without money? I mean he could delete your pc but idk... You seem to talk as if you already have capable AI systems. Mine is struggling to remember its todo list and to make basic systems that would allow it to work more efficiently. Granted the constant errors make it hard to say. Also the brilliance definitely shines trough some of the time. I think "AI" API that are save and where a AI can have something like a "credit card" for children where the parents have to confirm the purchase and so on and some things are restricted or go to the user for approval are nice. But i don't think we are there yet at all. Except computer safety if you have important stuff on your laptop or virtual machine environment or so

I agree that GPT is already extremely well censored (almost to a fault, in my opinion), but as I previously stated, there is already work in progress to make AutoGPT work with different, even local LLMs. In addition, bitcoin capabilities are included in the next PR batch. I don't think one can be too careful on these matters, and this isn't really even a tradeoff - it's an optional feature, after all.

File: .env.template
Comment (aslafy-z) [inline .env.template]: I believe this file should hold the same default than the ones in code. Either change that one or the other to align the values.

File: scripts/main.py
Comment (aslafy-z) [inline scripts/main.py]: I'm personally not into lower camel case flags. I'd prefer `--risk-avoidance`.

File: scripts/risk_evaluation.py
Comment (aslafy-z) [inline scripts/risk_evaluation.py]:
```suggestion
Your answers must be in the form `{"calculated_risk": <value between 0 and 1>, "reason": "<your reasons for calculating the value>"}`. Using the right syntax is vital to your task. Mind the curly brackets, the colon, the comma and the space. If you do not use the right syntax, you will be penalized.
```

Comment (aslafy-z) [reply scripts/risk_evaluation.py]: > you will be penalized

:laughing:

**Review by nponeccop — APPROVED** [review]

**Comment by LuposX** [comment]
I think It would be good to add examples to the prompt, for things that are risky and things which are not.

File: scripts/risk_evaluation.py
Comment (jxtrt) [reply scripts/risk_evaluation.py]: I also lol'd but gpt-3.5 tends to miss the prompt syntax way more often without that part

File: scripts/main.py
Comment (jxtrt) [reply scripts/main.py]: I don't have an opinion on this, whatever is agreed upon is fine by me.

File: .env.template
Comment (jxtrt) [reply .env.template]: Thank you, that was an oversight on my part. Fixed now.

**Comment by jnt0rrente** [comment]
> I think It would be good to add examples to the prompt, for things that are risky and things which are not.

I did attempt this approach, but I ultimately decided otherwise since I found omitting it provided good results, and so it would have been an unnecesary bias from my part. However, I would completely agree with adding a way for the user to provide their own opinion on risk, if and after this is merged.

**Review by nponeccop — DISMISSED** [review]

**Comment by nponeccop** [comment]
@Torantulino ?

**Comment by jnt0rrente** [comment]
Is there any more discussion needed? I'd like to resolve this and move on to other issues.

**Comment by nponeccop** [comment]
@richbeales @p-i- ?

**Comment by Pwuts** [comment]
@jnt0rrente are you planning to fix up this PR?

**Comment by jnt0rrente** [comment]
@Pwuts Yeah, upstream changes were way too big to merge. I'll rework this - hoping to merge it quicker than the first attempt.

**Comment by jnt0rrente** [comment]
@nponeccop @Torantulino Conflicts fixed, CI should be green.

**Comment by jnt0rrente** [comment]
Synced downstream and this closed autoaccidentally. Sorry guys - will reopen in a matter of minutes.

**Comment by jnt0rrente** [comment]
@nponeccop Are you planning on merging this any time soon?

File: AutoGpt.json
Comment (Pwuts) [inline AutoGpt.json:1]: This file shouldn't be committed

File: autogpt/cli.py
Comment (Pwuts) [inline autogpt/cli.py]: this is ugly, can it also work with `kebab-case`? (based on `--skip-reprompt` -> `skip_reprompt` probably yes)

**Review by Pwuts — DISMISSED** [review]
This is interesting. Team is looking into it.

File: autogpt/cli.py
Comment (jxtrt) [reply autogpt/cli.py]: Of course, thats completely my bad. I'll change it right now.

**Comment by jnt0rrente** [comment]
If you want my opinion, I think this should go into the base program rather than a plugin. It seems logical to me to offer the three options: assisted, full-auto and semi-auto - what this implements.

**Comment by jnt0rrente** [comment]
As it was said before: sure, GPT-n models are pretty much harmless and treat us all like 12-year-olds. Eventually, all of us will end up using a different, less censored model, and we will definitely feel safer knowing it won't start hallucinating and cause WWIII overnight.

**Comment by jnt0rrente** [comment]
@BillSchumacher any news?

**Comment by crushingCodes** [comment]
Im hoping this one gets merged too. I was thinking about adding something similar, except using a whitelist of commands I am ok with it running on its own. if this one does not work as expected i might still need to do that, but hoping this goes well!

**Comment by jnt0rrente** [comment]
@crushingCodes Yeah, I hope it is merged too but it seems like maintainers forgot about this PR. No news in a week.

**Comment by p-i-** [comment]
This is a mass message from the AutoGPT core team.
Our apologies for the ongoing delay in processing PRs.
This is because we are re-architecting the AutoGPT core!

For more details (and for infor on joining our Discord), please refer to:
  https://github.com/Significant-Gravitas/Auto-GPT/wiki/Architecting

**Comment by Boostrix** [comment]
this is heavily related to #2701

**Comment by Boostrix** [comment]
> If you want my opinion, I think this should go into the base program rather than a plugin. It seems logical to me to offer the three options: assisted, full-auto and semi-auto - what this implements.

That's a valid point actually - while this level of functionality could conceptually live inside a plugin, it would conceptually always be called by the core to work as it should, since it's a form of API injection anyway (IoC) - for pretty much the same reasons that constraint awareness (#3466) is a core feature, as exemplified by the ongoing work on implementing a dedicated budget manager implementation as part of the re-arch effort: #3919

**Comment by jnt0rrente** [comment]
I see this is getting traction again - I'm glad. If there is anything I can assist with, don't hesitate to @ me.

**Comment by ntindle** [comment]
If you wouldn't mind, can you check the tests and run the linting steps

**Comment by jnt0rrente** [comment]
> If you wouldn't mind, can you check the tests and run the linting steps

Will do, in about 2 hours.

**Comment by jnt0rrente** [comment]
@ntindle No idea why lint is complaining, checks pass on a fresh clone. 

In addition, I see run scripts (run.sh, run_continuous.sh) were added since I last checked. Should i add a run_riskavoidance.sh too? Idk what's your design idea.

File: .env.template
Comment (Pwuts) [inline .env.template:88]: Forgive me if this has been asked before, but why not use the smart model for this instead of having a separate setting for a risk eval model?

Comment (jxtrt) [reply .env.template:88]: I defined a new model parameter since gpt-3 was "good enough" in many cases, and it saves costs. It's possible, then, that the user wants to use the smart model for general thinking and the fast one for risk eval.

Comment (Pwuts) [reply .env.template:88]: Then the same question but with the fast model :)

Comment (jxtrt) [reply .env.template:88]: I most certainly don't get what you are asking. Are you suggesting we remove the choice entirely? I don't see the downside to allowing it.

Comment (Pwuts) [reply .env.template:88]: I am asking whether using the FAST_LLM_MODEL for risk eval will generally have the feature work as intended.

Having configuration parameters with overlapping purpose or meaning always increases the maintenance burden. So if we can avoid adding one, that would be nice.

Comment (jxtrt) [reply .env.template:88]: Well, I'd advise against that - gpt-3.5 can work well enough, but is generally stupider (in the sense that it has a harder time catching nuance, for example when evaluating shell commands) and more prone to extremes.

Comment (Pwuts) [reply .env.template:88]: Alright, thanks for elaborating!

File: autogpt/cli.py
Comment (Pwuts) [inline autogpt/cli.py:7]: `risk-avoidance` is technically accurate, although it doesn't reflect how this mode works, which is more of a supervisory workflow. Maybe something like `--self-supervise` would reflect it better. Or an option that combines risk avoidance and self-feedback:
* `--self-supervise=none` (default)
* `--self-supervise=guidance` for the existing self-feedback mode
* `--self-supervise=risk-averse` `--max-risk=0.3`

What do you think? cc @ntindle

**Review by Pwuts — COMMENTED** [review]
Because of naming, clarity and code structure, risk avoidance mode would be better as a modifier on top of continuous mode instead of a mode in itself. That way the whole thing becomes a bit more flexible so that it can possibly be combined with other modifiers in the future.

Similarly, `autogpt/risk_evaluation.py` could be `autogpt/self_evaluation.py`, which leaves room for other, similar functionality to be added later.

File: autogpt/cli.py
Comment (ntindle) [reply autogpt/cli.py:7]: I think pwuts’s suggestion is a really good way to handle it, especially in light of the work we are planningfor guardrails

**Comment by Boostrix** [comment]
> Similarly, `autogpt/risk_evaluation.py` could be `autogpt/self_evaluation.py`, which leaves room for other, similar functionality to be added later.

Agreed. 
See specifically:

- #4220 
- #4242

**Comment by eyalk11** [comment]
One should carefully think how to combine this with https://github.com/Significant-Gravitas/Auto-GPT/pull/3914.

**Pull request closed without merging**
